pkg/devfile/adapters/kubernetes/utils/utils.go

```diff
@@ -111,6 +112,27 @@ func GetContainers(devfileObj devfileParser.DevfileObj) ([]corev1.Container, err
 	return containers, nil
 }
 
+// GetEndpoints iterates through the components in the devfile and returns endpoints of all supported components
+func GetEndpoints(data data.DevfileData) (map[int32]common.Endpoint, error) {
+	endpointsMap := make(map[int32]common.Endpoint)
+
+	for _, comp := range adaptersCommon.GetDevfileContainerComponents(data) {
+		// Currently type container is the only devfile component that odo supports
+		if comp.Container.Endpoints != nil {
+			for _, endpoint := range comp.Container.Endpoints {
+				// TargetPort is a required entry for an Endpoint
+				// Devfile should not contains multiple identical TargetPorts, since all containers are inside one pod
+				if _, keyexist := endpointsMap[endpoint.TargetPort]; keyexist {
```

Comment:
Why are we using the `targetPort` as the key? Why not use the endpoint name?

Reply:
I totally agree with you that using endpoint name as the key makes more sense, if we has https://github.com/openshift/odo/issues/3718 implemented; which means for every entry in env.yaml we are going to have an exact match in devfile.yaml, then we can use the urlName as a key. 

For now, as described in https://github.com/openshift/odo/pull/3665#discussion_r468094439, we currently does not support duplication port in devfile, using `targetport` as a key is also unique.

Reply:
I have considered about this, so see the second point in issue description when I create this issue. 
https://github.com/openshift/odo/issues/3710#issue-674544229

```
update the logic of url creation. (i.e. any maps use port number as key should use URL name instead)
```

Reply:
@mik-dass Do you think that it would make sense to address this in separate PR as part of addressing https://github.com/openshift/odo/issues/3710?

Reply:
I am confused. If the user does `odo push` and the URL couldn't be created due to missing host information etc, what are we expecting/asking the user to do? 
- Create a URL with the exact name as defined in the endpoint?
- Create any URL on the same port and we will map it to the endpoint?

I like the first option better as it reduces a lot of confusion like https://github.com/openshift/odo/pull/3665#discussion_r468605247

> @mik-dass Do you think that it would make sense to address this in separate PR as part of addressing #3710?

@kadel Please have a look.

Reply:
See comment https://github.com/openshift/odo/pull/3665#discussion_r468640987

Reply:
https://github.com/openshift/odo/pull/3665#discussion_r468658887

Anyways leave it for now. Let's fix it in #3718 .